### PR TITLE
migrate to use k8s.io/util/net/port in kube-proxy 

### PR DESCRIPTION
**Pull request by JornShen**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

In proxier.go of iptables and ipvs, there are listenPortOpener with same name and same operation.
This MR use k8s.io/util/net/port to replace them. remove pkg/proxy/util/port and port_test.go

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
I think that this code is present in kubernetes/utils already, I didn´t check it in detail but IIRC the goal was to replace the kube-proxy one by the k/utils one
https://github.com/kubernetes/utils/blob/fddb29f9d0095a7a37582a1d86732ae83620f2e8/net/port.go#L44-L62

**Comment by JornShen** [comment]
> I think that this code is present in kubernetes/utils already, I didn´t check it in detail but IIRC the goal was to replace the kube-proxy one by the k/utils one
> https://github.com/kubernetes/utils/blob/fddb29f9d0095a7a37582a1d86732ae83620f2e8/net/port.go#L44-L62

@aojea 
you mean we should modify the k/util ? then kube-proxy use it?

**Comment by aojea** [comment]
> @aojea
> you mean we should modify the k/util ? then kube-proxy use it?

yeah, that was the original intent https://github.com/kubernetes/kubernetes/pull/80720#issuecomment-584636622

(why modify the k/util, is not valid right now?)

**Comment by JornShen** [comment]
> yeah, that was the original intent [#80720 (comment)](https://github.com/kubernetes/kubernetes/pull/80720#issuecomment-584636622)
> 
> (why modify the k/util, is not valid right now?)

valid, my fault :) thanks for guide 
I 'll use it to replace. :)
@aojea

File: pkg/proxy/util/utils_test.go [hunk 1]
```diff
@@ -927,5 +928,92 @@ func TestGetClusterIPByFamily(t *testing.T) {
...
+				{Port: 5004},
+			},
+			expectToBeClose: []bool{false, false, false},
+		},
+	}
 
+	for i, tc := range testCases {
+		replacementPortsMap := make(map[utilnet.LocalPort]utilnet.Closeable)
+		for _, lp := range tc.replacementPorts {
+			replacementPortsMap[lp] = &fakeClosable{}
```
Comment (aojea) [inline pkg/proxy/util/utils_test.go:1001]: is this imported from net/utils? it is failing in one of the jobs

Comment (JornShen) [reply pkg/proxy/util/utils_test.go:1010]: miss to migrate `fakeClosable`. 
updated :)
thanks !

**Comment by aojea** [comment]
I think that we have a "fortunate" coincidence with a flake test I was chasing for a long time, 
https://github.com/kubernetes/kubernetes/issues/91236
```
Kubernetes e2e suite: [sig-network] Conntrack should be able to preserve UDP traffic when server pod cycles for a NodePort service expand_less
```
seems that this PR always makes this test fail, it is likely that this is related, some race condition binding/forwarding?

**Comment by aojea** [comment]
if it fails 3/3 the same test it can be by chance :thinking: 
/test pull-kubernetes-e2e-gci-gce-ipvs
/test pull-kubernetes-e2e-gce-ubuntu-containerd

**Comment by JornShen** [comment]
> if it fails 3/3 the same test it can be by chance 🤔
> /test pull-kubernetes-e2e-gci-gce-ipvs
> /test pull-kubernetes-e2e-gce-ubuntu-containerd

@aojea 

still faild :(
I am seeing the it. Althougt I think this MR just use another pkg func to replace current, no change any code logic. A bit magical

**Comment by aojea** [comment]
> I am seeing the it. Althougt I think this MR just use another pkg func to replace current, no change any code logic. A bit magical

on the good side,  we have a lead to solve something that was flaking the CI for more than a year ;)

**Comment by aojea** [comment]
@JornShen can you add a temporal commit  with logging so we can use this PR to debug the problem?
So the test create 1 client pod, 1 server pod and a NodePort backed by the server port, something I will be interested to see in the logs is if the udp port that we are opening to hold the packet is receiving any packet, can you do that?

**Comment by aojea** [comment]
/hold
we are just debugging the problem :factory_worker:

File: test/e2e/network/conntrack.go [hunk 2]
```diff
@@ -78,13 +78,17 @@ var _ = SIGDescribe("Conntrack", func() {
...
 
-	logContainsFn := func(text, podName string) wait.ConditionFunc {
+	logContainsFn := func(text string, podName string, nodeIP string) wait.ConditionFunc {
 		return func() (bool, error) {
 			logs, err := e2epod.GetPodLogs(cs, ns, podName, podName)
 			if err != nil {
 				// Retry the error next time.
 				return false, nil
 			}
+			framework.Logf("Pod client logs: %s", logs)
```
Comment (aojea) [inline test/e2e/network/conntrack.go:88]: this is just for debugging, right?

Comment (JornShen) [reply test/e2e/network/conntrack.go:88]: yes

Comment (JornShen) [reply test/e2e/network/conntrack.go:88]: also ping below.

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -1100,17 +1096,18 @@ func (proxier *Proxier) syncProxyRules() {
 			// machine, hold the local port open so no other process can open it
 			// (because the socket might open but it would never work).
 			if (svcInfo.Protocol() != v1.ProtocolSCTP) && localAddrSet.Has(net.ParseIP(externalIP)) {
```
Comment (aojea) [inline pkg/proxy/iptables/proxier.go:1098]: just skip v1.ProtocolUDP too to not hold the local port, I want to discard that is this who is swallowing the packet

Comment (aojea) [reply pkg/proxy/iptables/proxier.go:1098]: sorry wong place, this is only for externalIPs

File: pkg/proxy/iptables/proxier.go [hunk 4]
```diff
@@ -1281,11 +1280,13 @@ func (proxier *Proxier) syncProxyRules() {
 					klog.V(4).InfoS("Port was open before and is still needed", "port", lp.String())
 					replacementPortsMap[lp] = proxier.portsMap[lp]
 				} else if svcInfo.Protocol() != v1.ProtocolSCTP {
```
Comment (aojea) [inline pkg/proxy/iptables/proxier.go:1282]: here
```suggestion
				} else if svcInfo.Protocol() == v1.ProtocolTCP {
```

File: pkg/proxy/ipvs/proxier.go [hunk 5]
```diff
@@ -1459,7 +1465,7 @@ func (proxier *Proxier) syncProxyRules() {
 					// We do not start listening on SCTP ports, according to our agreement in the
 					// SCTP support KEP
 				} else if svcInfo.Protocol() != v1.ProtocolSCTP {
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1467]:
```suggestion
				} else if svcInfo.Protocol() == v1.ProtocolTCP {
```

**Comment by aojea** [comment]
oh the ipvs job passed this time
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-gci-gce-ipvs

**Comment by aojea** [comment]
nevermind it keeps failing
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-gci-gce-ipvs

**Comment by aojea** [comment]
why is it working now? what has changed since last failure?
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-gci-gce-ipvs

File: pkg/proxy/iptables/proxier.go [hunk 6]
```diff
@@ -1281,12 +1280,14 @@ func (proxier *Proxier) syncProxyRules() {
...
-					socket, err := proxier.portMapper.OpenLocalPort(&lp, isIPv6)
+					socket, err := proxier.portMapper.OpenLocalPort(&lp)
 					if err != nil {
 						klog.ErrorS(err, "can't open port, skipping this nodePort", "port", lp.String())
 						continue
 					}
-					if lp.Protocol == "udp" {
+					klog.V(2).InfoS("Opened local port", "port", lp.String())
+
+					if lp.Protocol == utilnet.UDP {
```
Comment (aojea) [inline pkg/proxy/iptables/proxier.go:1290]: so, this was the problem, right?
we weren´t cleaning the conntrack entries and the test failed :thinking:

Comment (JornShen) [reply pkg/proxy/iptables/proxier.go:1290]: I think it cause that. The difference of them is here. The code before, this if judge dont take effect.

File: pkg/proxy/iptables/proxier.go [hunk 7]
```diff
@@ -1100,17 +1096,18 @@ func (proxier *Proxier) syncProxyRules() {
...
 			// (because the socket might open but it would never work).
 			if (svcInfo.Protocol() != v1.ProtocolSCTP) && localAddrSet.Has(net.ParseIP(externalIP)) {
-				lp := utilproxy.LocalPort{
+				lp := utilnet.LocalPort{
 					Description: "externalIP for " + svcNameString,
 					IP:          externalIP,
+					IPFamily:    localPortIPFamily,
 					Port:        svcInfo.Port(),
-					Protocol:    protocol,
+					Protocol:    utilnet.Protocol(svcInfo.Protocol()),
```
Comment (aojea) [inline pkg/proxy/iptables/proxier.go:1104]: this is unfortunate, so we are using lowercase in all the `proxy` code but in the library we use uppercase

Comment (JornShen) [reply pkg/proxy/iptables/proxier.go:1104]: That is the big differece between proxy/util/port.go and k8s.io/util/port.go. The other is LocalPort of k8s.io/util/port.go has new field ipfamily.

**Comment by JornShen** [comment]
@aojea can we cancel the hold ? and continue

**Comment by JornShen** [comment]
/cc @thockin
Could you also take a look?

File: pkg/proxy/ipvs/proxier.go (hunk 5 above)
Comment (aojea) [reply pkg/proxy/ipvs/proxier.go:1468]: discard this comment, it was for debugging

**Comment by aojea** [comment]
/hold cancel
Problem with the failing test was that we weren´t cleaning the conntrack entries,
https://github.com/kubernetes/kubernetes/pull/98130/files#r562508326

/lgtm
Thanks

**Comment by JornShen** [comment]
/cc @thockin 
Could you take a look?

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Comment by aojea** [comment]
/retest
lots of timeouts and 10 test failures that doesn't look related

**Comment by aojea** [comment]
/retest

Kubernetes e2e suite: [sig-network] Conntrack should drop INVALID conntrack entries expand_more

I think this flake will be fixed by https://github.com/kubernetes/kubernetes/pull/99147

**Pull request merged**